state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -42,84 +44,116 @@ pub const PALLET_VERSION: StorageVersion = StorageVersion::new(3);
 
 pub const SWAP_DELAY_BLOCKS: u32 = 2;
 
-#[derive(Clone, Debug, PartialEq, Eq, Encode, Decode, TypeInfo, MaxEncodedLen)]
-pub struct Swap {
-	swap_id: SwapId,
-	pub from: Asset,
-	pub to: Asset,
-	input_amount: AssetAmount,
-	swap_type: SwapType,
+struct SwapCtx {
+	swap: Swap,
 	stable_amount: Option<AssetAmount>,
 	final_output: Option<AssetAmount>,
 }
 
-#[derive(Clone, Debug, PartialEq, Eq, Encode, Decode, TypeInfo, MaxEncodedLen)]
-pub struct SwapLegInfo {
-	pub swap_id: SwapId,
-	pub base_asset: Asset,
-	pub quote_asset: Asset,
-	pub side: Side,
-	pub amount: AssetAmount,
-	pub source_asset: Option<Asset>,
-	pub source_amount: Option<AssetAmount>,
-}
-
-impl Swap {
-	fn new(
-		swap_id: SwapId,
-		from: Asset,
-		to: Asset,
-		input_amount: AssetAmount,
-		swap_type: SwapType,
-	) -> Self {
+impl SwapCtx {
+	fn new(swap: Swap) -> Self {
 		Self {
-			swap_id,
-			from,
-			to,
-			input_amount,
-			swap_type,
-			stable_amount: if from == STABLE_ASSET { Some(input_amount) } else { None },
-			final_output: if from == to { Some(input_amount) } else { None },
+			stable_amount: if swap.from == STABLE_ASSET { Some(swap.input_amount) } else { None },
+			final_output: if swap.from == swap.to { Some(swap.input_amount) } else { None },
+			swap,
 		}
 	}
 
-	fn swap_asset(&self, direction: SwapLeg) -> Option<Asset> {
-		match (direction, self.from, self.to) {
-			(SwapLeg::ToStable, STABLE_ASSET, _) => None,
-			(SwapLeg::ToStable, from, _) => Some(from),
-			(SwapLeg::FromStable, _, STABLE_ASSET) => None,
-			(SwapLeg::FromStable, _, to) => Some(to),
-		}
+	fn swap_id(&self) -> SwapId {
+		self.swap.swap_id
 	}
 
-	fn swap_amount(&self, direction: SwapLeg) -> Option<AssetAmount> {
-		match direction {
-			SwapLeg::ToStable => Some(self.input_amount),
-			SwapLeg::FromStable => self.stable_amount,
-		}
+	fn input_asset(&self) -> Asset {
+		self.swap.from
```

Comment:
nit: the value is public, we could just rename it. (or make it private if the intent here is to prevent mutating the value)

Reply:
The intent is to avoid reaching for a nested value through the inner swap and make the higher level code a bit nicer. I kept these public because they are needed elsehwere (in the swap subscription).